### DISCUSSION TITLE
Parsing a Float out of a String

### DISCUSSION THREAD

**Discussion by darrenjw** [google-research/dex-lang#973] [Q&A] [answered]
For getting data into Dex it would be useful to be able to parse a Float out of a String. I guess I'm looking for a function with a signature like
```
parse_float (s: String) : Maybe Float
```
or anything vaguely like "read" in Haskell or "fscanf" in C. Obviously the compiler knows how to do this, but I haven't managed to find it surfaced in Dex. Is there anything like this that I'm missing?

**Comment by chenzizhao** [comment]
@cyntsh figured out [`FilePath -> List Float`](https://github.com/google-research/dex-lang/pull/546/files#diff-5c0cec0e362265f4f7ccac6cae0fbd4d875ab7a2611d37e9451715d6ff93cd7eR72-R78) which seems similar to what you were looking for. #543 might be relevant too.

**Reply by darrenjw** [reply]
Thanks, but I think both of these are related to casting floats stored in binary formats into Dex Floats. What I am interested in is turning a String representation of a Float into a float. So, eg. the ASCII String "12.34" will get turned into the Dex Float 12.34.

**Comment by apaszke** [comment, marked as answer]
Great question! I think this is one of the pieces we're missing at the moment. Nothing prevents us from reimplementing the `Read` typeclass in Dex, but it's code that has to be written. I might try to put something together today.

In general, we've been talking for a while about writing a CSV library for Dex, and I imagine this would be one of the prerequisites for that.

**Reply by apaszke** [reply]
Turns out this can be done fairly easily through C FFI. I wrote an implementation for `Float` in https://github.com/google-research/dex-lang/pull/974. Please let us know if you need it for any other types!

**Reply by darrenjw** [reply]
Very cool - many thanks!
